pkg/utils/csr_helpers.go

```diff
@@ -0,0 +1,144 @@
+package utils
+
+import (
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/cloudflare/cfssl/config"
+	"github.com/cloudflare/cfssl/signer"
+	"github.com/cloudflare/cfssl/signer/local"
+	certificatesv1 "k8s.io/api/certificates/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog/v2"
+	"open-cluster-management.io/addon-framework/pkg/agent"
+	addonapiv1alpha1 "open-cluster-management.io/api/addon/v1alpha1"
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+)
+
+// DefaultSigner generates a signer func for addon agent to sign the csr using caKey and caData.
+func DefaultSigner(caKey, caData []byte) agent.CSRSignerFunc {
+	return func(csr *certificatesv1.CertificateSigningRequest) []byte {
+		blockTlsCrt, _ := pem.Decode(caData) // note: the second return value is not error for pem.Decode; it's ok to omit it.
+		certs, err := x509.ParseCertificates(blockTlsCrt.Bytes)
+		if err != nil {
+			klog.Errorf("Failed to parse cert: %v", err)
+			return nil
+		}
+
+		blockTlsKey, _ := pem.Decode(caKey)
+		key, err := x509.ParsePKCS1PrivateKey(blockTlsKey.Bytes)
+		if err != nil {
+			klog.Errorf("Failed to parse key: %v", err)
+			return nil
+		}
+
+		data, err := signCSR(csr, certs[0], key)
+		if err != nil {
+			klog.Errorf("Failed to sign csr: %v", err)
+			return nil
+		}
+		return data
+	}
+}
+
+func signCSR(csr *certificatesv1.CertificateSigningRequest, caCert *x509.Certificate, caKey *rsa.PrivateKey) ([]byte, error) {
+	var usages []string
+	for _, usage := range csr.Spec.Usages {
+		usages = append(usages, string(usage))
+	}
+
+	certExpiryDuration := 365 * 24 * time.Hour
```

Comment:
how about adding another public method `DefaultSignerWithExpiry` that allows custom expiry duration?

Reply:
yes, just change to DefaultSignerWithExpiry.